srcpkgs/cmt/template

```diff
@@ -12,7 +12,7 @@ maintainer="Olga Ustuzhanina <me@laserbat.pw>"
 license="GPL-2.0-or-later"
 homepage="http://www.ladspa.org/cmt/overview.html"
 distfiles="http://www.ladspa.org/download/${pkgname}_${version}.tgz"
-checksum=eb56d7abebfdf8a6d0ad65d012238c9fc394dd41eeca11900812a8cb6b07ad1f
+checksum=a82f8636de1f4ada386a199a017a9cd775a49b49e716b11e8dd3f723c93df6ca
 
 post_extract() {
 	sed -e "/^CFLAGS/ s/-O2/${CFLAGS}/" \
```

Comment:
I don't think `sed` would be even necessary when `make_use_env` wouldn't be set. Xbps will override it whether the lines are there or not (when `make_use_env` is unset). But both of these approaches (just unsetting `make_use_env` or unsetting `make_use_env` and then doing `sed -e "/^CFLAGS/d; /^CXXFLAGS/d" -i "${build_wrksrc}/Makefile"`) have a flaw: they remove the `-fPIC` flag from `CFLAGS`. This flag affects compilation in a major way and shouldn't be removed without reason. `-fPIC` could be added to `make_build_args`. What do You think?

Reply:
My apologies for the delay.

> I don't think sed would even be necessary when make_use_env wouldn't be set.

cmt's makefile defines CFLAGS and CXXFLAGS with `=` which completely overrides external variables as far as I can tell.

Here is what the makefile contains:
```
CFLAGS		=	$(INCLUDES) -Wall -Werror -O2 -fPIC
CXXFLAGS	=	$(CFLAGS)
PLUGIN_LIB	=	../plugins/cmt.so
```

> But both of these approaches have a flaw: they remove the `-fPIC` flag from `CFLAGS`.

That would be my fault for overlooking this.

Using sed to stuff it with void's flags just because `-fPIC` is present there looks overcomplicated and unnecessary.

In this case I can suggest the following (I forgot `vsed` must be used instead of `sed`):

```diff
diff --git a/srcpkgs/cmt/template b/srcpkgs/cmt/template
index a07ca8a069..6c676b62b9 100644
--- a/srcpkgs/cmt/template
+++ b/srcpkgs/cmt/template
@@ -14,9 +14,9 @@ distfiles="https://www.ladspa.org/download/${pkgname}_${version}.tgz"
 checksum=a82f8636de1f4ada386a199a017a9cd775a49b49e716b11e8dd3f723c93df6ca
 
 post_extract() {
-	sed -e "/^CFLAGS/ s/-O2/${CFLAGS}/" \
+	vsed -e "s,^\(CFLAGS.*\)=,\1+=," \
 	  -e 's|-Werror||g' \
-	  -i "${wrksrc}/src/Makefile"
+	  -i "${build_wrksrc}/Makefile"
 }
 
 do_install() {
```

`=` is substituted with `+=`, thus making externally defined `CFLAGS` and `CXXFLAGS` usable.

`make_use_env` now has its use here unlike earlier:
```markdown
This build style tries to compensate for makefiles
that do not respect environment variables, so well written makefiles, those
that do such things as append (`+=`) to variables, should have `make_use_env`
set in the body of the template.
```

With this approach only two lines have to be changed without meddling with upstream flags.

> `-fPIC` could be added to `make_build_args`

Based on my previous answer (I overlooked this flag for reasons unknown), I believe it should be left in the makefile since the fix above allows to use void's flags with those in the makefile together without that ugly hack.

Reply:
> cmt's makefile defines CFLAGS and CXXFLAGS with `=` which completely overrides external variables as far as I can tell.

When `$make_use_env` is unset, xbps-src [very explicitly](https://github.com/void-linux/void-packages/blob/master/common/build-style/gnu-makefile.sh#L7-L13) overrides all relevant variables set in the makefile. It doesn't matter what cmt's makefile wants to do, xbps-src will set it (unless it would be using the `override` keyword or other shenanigans but no one does that for `CFLAGS`).

But as I said it would still omit `-fPIC` which is undesirable.

> Using sed to stuff it with void's flags just because `-fPIC` is present there looks overcomplicated and unnecessary.
> 
> In this case I can suggest the following (I forgot `vsed` must be used instead of `sed`):

I agree that it looks overcomplicated. I think using sed in general isn't really fit for this situation. I'll make an actual patch that fixes this.

While I'm at it I'll try to contact upstream about this.

> > `-fPIC` could be added to `make_build_args`
> 
> Based on my previous answer (I have had overlooked this flag for reasons unknown), I believe it should be left in the makefile since the fix above allows to use void's flags with those in the makefile together without that ugly hack.

This is the best solution. Template files shouldn't be responsible for this.

I have experience with writing makefiles. The absolute best solution is https://www.gnu.org/software/make/manual/make.html#Command-Variables or (my favorite) presume that `CFLAGS` and similar variables are user input and put mandatory flags right into the rule from the beginning like this:
```makefile
a.o : a.c b.c
	$(CC) -c -fPIC --you-can't-override-me-when-I'm-hardcoded=true $(CFLAGS) $(CPPFLAGS) a.c b.c -o a.o
```
but it isn't xbps-src's responsibility to fix build system of other programs and both of these changes would require large patches so I'll make something simpler (probably just change `=` to `+=` but I'm not sure yet). I will propose one of the "correct" solutions to upstream.

Reply:
> probably just change `=` to `+=` but I'm not sure yet

That would be for the best *if* the upstream proves to be not amenable to your proposal.